pkg/reconciler/pipelinerun/controller.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2021 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"context"
+
+	"github.com/tektoncd/chains/pkg/chains"
+	"github.com/tektoncd/chains/pkg/chains/storage"
+	"github.com/tektoncd/chains/pkg/config"
+	pipelineclient "github.com/tektoncd/pipeline/pkg/client/injection/client"
+	pipelineruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1beta1/pipelinerun"
+	pipelinerunreconciler "github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/pipelinerun"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	logger := logging.FromContext(ctx)
+	pipelineRunInformer := pipelineruninformer.Get(ctx)
+
+	kubeClient := kubeclient.Get(ctx)
+	pipelineClient := pipelineclient.Get(ctx)
+
+	psSigner := &chains.ObjectSigner{
+		SecretPath:        SecretPath,
+		Pipelineclientset: pipelineClient,
+	}
+
+	c := &Reconciler{
+		PipelineRunSigner: psSigner,
+		Pipelineclientset: pipelineClient,
+	}
+	impl := pipelinerunreconciler.NewImpl(ctx, c, func(impl *controller.Impl) controller.Options {
+		cfgStore := config.NewConfigStore(logger, func(name string, value interface{}) {
+			// get updated config
+			cfg := *value.(*config.Config)
+
+			// get all formatters for formatting payload
+			psSigner.Formatters = chains.AllFormatters(cfg, logger)
+
+			// get all backends for storing provenance
+			backends, err := storage.InitializeBackends(ctx, pipelineClient, kubeClient, logger, cfg)
```

Comment:
If you want to restrict which backends are available for which type, this is a good place to do it.

You're loading all backends for pipelineruns here, even though you call out several of them as TaskRun only above.

Reply:
Would [here](https://github.com/hacbs-contract/chains/blob/poc-tep-84/pkg/config/config.go#L249) be a better place? This way the deployment fails right away due to misconfiguration.

Reply:
Ended up doing this during the initial configuration check.